### PR TITLE
JAMES-3516 Create SearchGuessingAlgorithm and inject it into ThreadIdGuessingAlgorithm

### PR DESCRIPTION
**Pull request by quantranhong1999**
My idea is using MessageSearchIndex to get messageIds of a user, then use MessageIdMapper to find Mailbox Messages based on messageIds.

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SearchGuessingAlgorithmImpl.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        return searchMessageIds(expression, session, limit)
+            .flatMap(messageId -> messageIdMapper.findReactive(Collections.singletonList(messageId), fetchType));
+    }
+
+    @Override
+    public Flux<MessageId> searchMessageIds(MultimailboxesSearchQuery expression, MailboxSession session, long limit) throws MailboxException {
+        return getInMailboxIds(expression, session)
+            .filter(id -> !expression.getNotInMailboxes().contains(id))
+            .collect(Guavate.toImmutableSet())
+            .flatMapMany(Throwing.function(ids -> index.search(session, ids, expression.getSearchQuery(), limit)));
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SearchGuessingAlgorithmImpl.java:71]: Use mailboxManager for the search...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SearchGuessingAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.MultimailboxesSearchQuery;
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+
+import reactor.core.publisher.Flux;
+
+public interface SearchGuessingAlgorithm {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SearchGuessingAlgorithm.java:31]: I though implementing ThreadIdGuessingAlgorithm...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/NaiveThreadIdGuessingAlgorithmImpl.java [hunk 3]
```diff
@@ -22,15 +22,38 @@
...
+import org.apache.james.mailbox.model.SearchQuery;
 import org.apache.james.mailbox.model.ThreadId;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
 import org.apache.james.mailbox.store.mail.model.MimeMessageId;
 import org.apache.james.mailbox.store.mail.model.Subject;
+import org.apache.james.mailbox.store.search.SearchGuessingAlgorithm;
+
+import reactor.core.publisher.Flux;
 
 public class NaiveThreadIdGuessingAlgorithmImpl implements ThreadIdGuessingAlgorithm {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/NaiveThreadIdGuessingAlgorithmImpl.java:41]: No keep this class unchanged and implement another class for finding the threadId by doing searches...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SearchGuessingAlgorithmImpl.java (hunk 1 above)
Comment (quantranhong1999) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SearchGuessingAlgorithmImpl.java:71]: That would be a loop: MailboxManager need ThreadIdGuessingAlgorithm and ThreadIdGuessingAlgorithm need MailboxManager

Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SearchGuessingAlgorithmImpl.java:71]: Injection systems like guice are designed to break these loops.

**Pull request closed without merging**